### ISSUE TITLE
Optimized `flip_normals` changes downstream operations

### ISSUE THREAD

**Issue by whophil** [pyvista/pyvista#1669]
**Describe the bug, what's wrong, and what you expect:**

#1602 has changed the behavior of operations _downstream_ of `PolyData.flip_normals()`. The output of `flip_normals()` on the data object itself appears unchanged.

-----

**To Reproduce**
```py
import pyvista as pv
import numpy as np

faces = [[3, 0, 1, 3], [3, 0, 3, 2], [3, 2, 3, 5], [3, 5, 4, 2]]
points = [[1., 0., 0.], [0., 0., 0.], [1., 1., 2.], [0., 1., 2.], [1., 3., 1.], [0., 3., 1.]]

mesh = pv.PolyData(points, faces)
mirror = mesh.reflect((0., 1., 0.))
mirror.flip_normals()
print(np.reshape(mirror.faces, (-1, 4)))

full = mesh + mirror
print(np.reshape(full.faces, (-1, 4)))

```

Using the legacy version of `flip_normals()` (as in PyVista 0.31), this produces:
```
[[3 3 1 0]
 [3 2 3 0]
 [3 5 3 2]
 [3 2 4 5]]

[[3 0 1 2]
 [3 0 2 3]
 [3 3 2 4]
 [3 4 5 3]
 [3 6 1 0]
 [3 7 6 0]
 [3 8 6 7]
 [3 7 9 8]]
```

Using the optimized version of `flip_normals()` (as in PyVista 0.32), this produces:
```
[[3 3 1 0]
 [3 2 3 0]
 [3 5 3 2]
 [3 2 4 5]]

[[3 0 1 2]
 [3 3 0 2] <-- Differences start here!
 [3 4 0 3]
 [3 3 5 4]
 [3 6 1 2]
 [3 7 6 2]
 [3 8 6 7]
 [3 7 9 8]]
```
To check that these changes are indeed due to changes in `flip_normals()`, these outputs were generated by pasting in the legacy version of the `flip_normals()` routine into `poly_data.py` instead of by changing PyVista versions.

-----

**System Information:**
 Please run the following code wherever you are experiencing the bug and paste the output below. This report helps us track down bugs and it is critical to addressing your bug:

```txt
--------------------------------------------------------------------------------
  Date: Tue Sep 14 09:06:17 2021 MDT
                OS : Linux
            CPU(s) : 16
           Machine : x86_64
      Architecture : 64bit
       Environment : IPython
        GPU Vendor : NVIDIA Corporation
      GPU Renderer : Quadro T2000 with Max-Q Design/PCIe/SSE2
       GPU Version : 4.5.0 NVIDIA 460.91.03
  Python 3.8.11 (default, Aug  3 2021, 15:09:35)  [GCC 7.5.0]
           pyvista : 0.32.1
               vtk : 9.0.1
             numpy : 1.21.2
           imageio : 2.9.0
           appdirs : 1.4.4
            scooby : 0.5.7
            meshio : 4.4.6
        matplotlib : 3.4.3
             PyQt5 : 5.12.3
           IPython : 7.27.0
             scipy : 1.7.1
--------------------------------------------------------------------------------
```

**Comment by adeak** [comment]
I can confirm what you see between before and after the linked PR. Thanks for the great example.

Printing `full.points` gives me on the newer version
```
[[ 0.  1.  2.]
 [ 0.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  1.  2.]
 [ 0.  3.  1.]
 [ 1.  3.  1.]
 [ 0. -1.  2.]
 [ 1. -1.  2.]
 [ 0. -3.  1.]
 [ 1. -3.  1.]]
```

While on the older one
```
[[ 1.  0.  0.]
 [ 0.  0.  0.]
 [ 0.  1.  2.]
 [ 1.  1.  2.]
 [ 0.  3.  1.]
 [ 1.  3.  1.]
 [ 0. -1.  2.]
 [ 1. -1.  2.]
 [ 0. -3.  1.]
 [ 1. -3.  1.]]
```

Since points are ordered differently, the cells will have different point indices even if they are the same. I haven't checked yet whether the cells do indeed correspond to one another.

Do you have code relying on the order or labelling of cells?

**Comment by whophil** [comment]
@adeak thanks for looking into this. I think there is more to it than just the point ordering (but I agree that the point ordering shouldn't change due to the PR).

Here is a slightly modified example of the code above which:
1. modifies the mesh to sit on the z-axis
2. computes the signed area as `cross(a, b)` where `a = p1 - p0` and `b = p2 - p0` of any face

The output shows that the signed area (which is representative of the normal direction for a triangular cell) is actually changing due to the optimization of `flip_normals`

With legacy flip_normals
```
[[3 3 1 0]
 [3 2 3 0]
 [3 5 3 2]
 [3 2 4 5]]
[[3 0 1 2]
 [3 0 2 3]
 [3 3 2 4]
 [3 4 5 3]
 [3 6 1 0]
 [3 7 6 0]
 [3 8 6 7]
 [3 7 9 8]]
[-0.5, -0.5, -0.5, -0.5, -0.5, -0.5, -0.5, -0.5]
```

With optimized flip_normals
```
[[3 3 1 0]
 [3 2 3 0]
 [3 5 3 2]
 [3 2 4 5]]
[[3 0 1 2]
 [3 3 0 2]
 [3 4 0 3]
 [3 3 5 4]
 [3 6 1 2]
 [3 7 6 2]
 [3 8 6 7]
 [3 7 9 8]]
[0.5, 0.5, 0.5, 0.5, -0.5, -0.5, -0.5, -0.5]
```

**Comment by adeak** [comment]
@whophil excellent, thanks. Could you please post your updated code? 

The sign difference in the normal direction _is_ concerning, because the whole point of the `flip_normals()` filter is to reverse the order of points in each triangle, thereby changing sign. So it would also be interesting to compare the normal directions _before_ calling `flip_normals()`. Either the difference is there to begin with (which would be surprising), or it wasn't (in which case `flip_normals()` is flat-out buggy, because it doesn't flip all normals for some reason).

**Comment by adeak** [comment]
OK, I think this is the problem:
```py
import pyvista as pv

faces = [[3, 0, 1, 3], [3, 0, 3, 2], [3, 2, 3, 5], [3, 5, 4, 2]]
points = [[1., 0., 0.], [0., 0., 0.], [1., 1., 2.], [0., 1., 2.], [1., 3., 1.], [0., 3., 1.]]

mesh = pv.PolyData(points, faces)
mirror = mesh.reflect((0., 1., 0.))
mirror.flip_normals()
print(mesh.compute_normals()['Normals'])
```

Old version:
```
[[ 0.         0.8944272 -0.4472136]
 [ 0.         0.8944272 -0.4472136]
 [ 0.        -0.4472136 -0.8944272]
 [-0.        -0.4472136 -0.8944272]]
```

New version:
```
[[ 0.        -0.8944272  0.4472136]
 [ 0.        -0.8944272  0.4472136]
 [-0.         0.4472136  0.8944272]
 [ 0.         0.4472136  0.8944272]]
```

For some reason calling `flip_normals()` on `mesh.reflect(...)` also flips the normals on the original mesh. This sounds like the faces array being shared between `mesh` and `mirror` in the above example... which is not great!

The new VTK9 API exposes shallow copies that weren't visible earlier:
```py
>>> import pyvista as pv
>>> mesh = pv.PolyData()
>>> mesh.faces = [1, 42, 2, -1, 25]
>>> other = mesh.reflect((0, 0, 1))
>>> mesh.faces
array([ 1, 42,  2, -1, 25])
>>> other.faces
array([ 1, 42,  2, -1, 25])
>>> other.faces[:] = [2, 3, 4, 1, 5]  # doesn't affect the original
>>> mesh.faces
array([ 1, 42,  2, -1, 25])
>>> other._connectivity_array[:] = [3, 4, 5]  # affects the original
>>> mesh.faces
array([1, 3, 2, 4, 5])
```
Sounds like we should yet again apply generous deep copies in the library... (or find the potential bug in `reflect()`).
cc @akaszynski

**Comment by whophil** [comment]
@adeak thanks very much for digging into this. The `reflect()` filter just delegates to `vtkTransformFilter` eventually. This may be of interest...

```py
import pyvista as pv
import numpy as np

faces = [[3, 0, 1, 3], [3, 0, 3, 2], [3, 2, 3, 5], [3, 5, 4, 2]]
points = [[1., 0., 0.], [0., 0., 0.], [1., 1., 0.], [0., 1., 0.], [1., 2., 0.], [0., 2., 0.]]

mesh = pv.PolyData(points, faces)

t = np.array(np.eye(4))
mesh2 = mesh.transform(t, inplace=False)

print(np.shares_memory(mesh2.faces, mesh.faces))
print(np.shares_memory(mesh2.points, mesh.points))
```

Gives
```
True
False
```

**Comment by adeak** [comment]
Hmm, but this is not what I'd have expected...
```py
>>> mesh.faces.base is mesh2.faces.base
True
>>> mesh._connectivity_array.base is mesh2._connectivity_array.base
True
```
This suggests that the `.faces` array also views the same underlying data... which is not what my earlier mutation attempt and the whole situation suggests.

In any case I've reached the border of my ignorance concerning VTK, so we'll need @akaszynski or someone else to go further and patch this up.

**Comment by akaszynski** [comment]
@adeak is correct.  Smoking gun is here:

```py
import pyvista as pv
import numpy as np

faces = [[3, 0, 1, 3], [3, 0, 3, 2], [3, 2, 3, 5], [3, 5, 4, 2]]
points = [[1., 0., 0.], [0., 0., 0.], [1., 1., 2.], [0., 1., 2.], [1., 3., 1.], [0., 3., 1.]]

mesh = pv.PolyData(points, faces)

trans = np.diag(np.ones(4))
trans[:, -1] = 1

transformed = mesh.transform(trans, inplace=False)

mesh._connectivity_array[0] += 1
np.array_equal(mesh._connectivity_array, transformed._connectivity_array)
```

In VTK 9, the `faces` attribute is not a view on the underlying connectivity array, it's a derived array there for legacy purposes.

---

What we've discovered by updating the connectivity arrays in situ with `flip_normals`, is the connectivity array is a shallow copy of the original array regardless of the setting of `inplace`.  This is a bug and I'll fix it in a follow up PR.

**Comment by akaszynski** [comment]
This has been resolved by #1688 and I'm unable to reproduce this issue on my end.  @whophil, please let me know if this is still an issue and I'll reopen.
